streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -365,10 +370,25 @@ synchronized boolean lock(final TaskId taskId) {
             lockedTasksToOwner.put(taskId, Thread.currentThread());
             // make sure the task directory actually exists, and create it if not
             getOrCreateDirectoryForTask(taskId);
+            lockedTasksToBackoffRecord.remove(taskId);
             return true;
         }
     }
 
+    public boolean canTryLock(final TaskId taskId, final long nowMs) {
+        return !lockedTasksToBackoffRecord.containsKey(taskId) || lockedTasksToBackoffRecord.get(taskId).canAttempt(nowMs);
+    }
+
+    private void updateOrCreateBackoffRecord(final TaskId taskId, final long nowMs) {
+        if (lockedTasksToBackoffRecord.containsKey(taskId)) {
+            final BackoffRecord backoffRecord = lockedTasksToBackoffRecord.get(taskId);
+            backoffRecord.lastAttemptMs = nowMs;
```

Comment:
nit: maybe we could move the update of the fields also into a method of BackoffRecord, something like backoffRecord.recordAttempt(nowMs); ?